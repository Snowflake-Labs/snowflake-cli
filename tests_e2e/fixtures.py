import shutil
from contextlib import contextmanager
from pathlib import Path

import pytest


@pytest.fixture
def project_directory(temp_dir, test_root_path):
    @contextmanager
    def _temporary_project_directory(project_name):
        test_data_file = test_root_path / "test_data" / project_name
        shutil.copytree(test_data_file, temp_dir, dirs_exist_ok=True)
        yield Path(temp_dir)

    return _temporary_project_directory
