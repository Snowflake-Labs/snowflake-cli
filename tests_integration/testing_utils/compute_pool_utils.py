import math
import time

import pytest
from snowflake.connector import SnowflakeConnection

from tests_integration.conftest import SnowCLIRunner
from tests_integration.test_utils import contains_row_with, not_contains_row_with
from tests_integration.testing_utils.assertions.test_result_assertions import (
    assert_that_result_is_successful_and_executed_successfully,
)


class ComputePoolTestSetup:
    def __init__(
        self,
        runner: SnowCLIRunner,
        snowflake_session: SnowflakeConnection,
    ):
        self.runner = runner
        self.snowflake_session = snowflake_session


class ComputePoolTestSteps:
    def __init__(self, setup: ComputePoolTestSetup):
        self._setup = setup

    def create_compute_pool(self, compute_pool_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "spcs",
                "compute-pool",
                "create",
                compute_pool_name,
                "--min-nodes",
                1,
                "--family",
                "CPU_X64_XS",
            ]
        )
        assert result.json, result.output
        assert "status" in result.json
        assert (
            f"Compute Pool {compute_pool_name.upper()} successfully created."
            in result.json["status"]  # type: ignore
        )

    def list_should_return_compute_pool(self, compute_pool_name) -> None:
        result = self._execute_list()
        assert contains_row_with(result.json, {"name": compute_pool_name.upper()})

    def list_should_not_return_compute_pool(self, compute_pool_name: str) -> None:
        result = self._execute_list()
        assert not_contains_row_with(result.json, {"name": compute_pool_name.upper()})

    def describe_should_return_compute_pool(self, compute_pool_name: str) -> None:
        result = self._execute_describe(compute_pool_name)
        assert result.json
        assert len(result.json) == 1
        assert result.json[0]["name"] == compute_pool_name.upper()

    def drop_compute_pool(self, compute_pool_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "object",
                "drop",
                "compute-pool",
                compute_pool_name,
            ],
        )
        assert result.json
        assert len(result.json) == 1
        assert result.json[0] == {  # type: ignore
            "status": f"{compute_pool_name.upper()} successfully dropped."
        }

    def stop_all_on_compute_pool(self, compute_pool_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            ["spcs", "compute-pool", "stop-all", compute_pool_name]
        )
        assert_that_result_is_successful_and_executed_successfully(result, is_json=True)

    def suspend_compute_pool(self, compute_pool_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            ["spcs", "compute-pool", "suspend", compute_pool_name]
        )
        assert_that_result_is_successful_and_executed_successfully(result, is_json=True)

    def resume_compute_pool(self, compute_pool_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            ["spcs", "compute-pool", "resume", compute_pool_name]
        )
        assert_that_result_is_successful_and_executed_successfully(result, is_json=True)

    def wait_until_compute_pool_is_idle(self, compute_pool_name: str) -> None:
        self._wait_until_compute_pool_reaches_state(compute_pool_name, "IDLE", 300)

    def wait_until_compute_pool_is_suspended(self, compute_pool_name: str) -> None:
        self._wait_until_compute_pool_reaches_state(compute_pool_name, "SUSPENDED", 60)

    def _wait_until_compute_pool_reaches_state(
        self, compute_pool_name: str, target_state: str, max_duration: int
    ):
        assert max_duration > 0
        max_counter = math.ceil(max_duration / 10)
        target_state = target_state.upper()
        for i in range(max_counter):
            status = self._execute_describe(compute_pool_name)
            if contains_row_with(status.json, {"state": target_state}):
                return
            time.sleep(10)
        status = self._execute_describe(compute_pool_name)

        error_message = f"Compute pool {compute_pool_name} didn't reach target state '{target_state}' in {max_duration} seconds. Current state is '{status.json['state']}'"
        pytest.fail(error_message)

    def _execute_describe(self, compute_pool_name: str):
        return self._setup.runner.invoke_with_connection_json(
            ["object", "describe", "compute-pool", compute_pool_name]
        )

    def _execute_list(self):
        return self._setup.runner.invoke_with_connection_json(
            [
                "object",
                "list",
                "compute-pool",
            ],
        )
