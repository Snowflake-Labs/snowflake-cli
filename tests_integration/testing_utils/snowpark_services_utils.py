import json
import math
import time
from textwrap import dedent
from typing import Union

import pytest
from snowflake.connector import SnowflakeConnection

from tests_integration.conftest import SnowCLIRunner
from tests_integration.test_utils import contains_row_with, not_contains_row_with
from tests_integration.testing_utils.assertions.test_result_assertions import (
    assert_that_result_is_successful_and_output_json_contains,
    assert_that_result_is_successful_and_output_json_equals,
)


class SnowparkServicesTestSetup:
    def __init__(
        self,
        runner: SnowCLIRunner,
        snowflake_session: SnowflakeConnection,
        test_root_path,
    ):
        self.runner = runner
        self.snowflake_session = snowflake_session
        self.test_root_path = test_root_path


class SnowparkServicesTestSteps:
    compute_pool = "snowcli_compute_pool"
    database = "snowcli_db"
    schema = "public"
    container_name = "echo-test"

    def __init__(self, setup: SnowparkServicesTestSetup):
        self._setup = setup

    def create_service(self, service_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "spcs",
                "service",
                "create",
                service_name,
                "--compute-pool",
                self.compute_pool,
                "--spec-path",
                self._get_spec_path(),
                "--database",
                self.database,
                "--schema",
                self.schema,
            ],
        )
        assert_that_result_is_successful_and_output_json_equals(
            result, {"status": f"Service {service_name.upper()} successfully created."}
        )

    def status_should_return_service(self, service_name: str) -> None:
        result = self._execute_status(service_name)
        assert_that_result_is_successful_and_output_json_contains(
            result,
            {"containerName": self.container_name, "serviceName": service_name.upper()},
        )

    def logs_should_return_service_logs(self, service_name: str) -> None:
        result = self._setup.runner.invoke_with_connection(
            [
                "spcs",
                "service",
                "logs",
                service_name,
                "--container-name",
                self.container_name,
                "--instance-id",
                "0",
                "--database",
                self.database,
                "--schema",
                self.schema,
            ],
        )
        assert result.output
        # Assert this instead of full payload due to log coloring
        assert service_name in result.output
        assert '"GET /healthcheck HTTP/1.1" 200 -' in result.output

    def list_should_return_service(self, service_name: str) -> None:
        result = self._execute_list()
        assert contains_row_with(result.json, {"name": service_name.upper()})

    def list_should_not_return_service(self, service_name: str) -> None:
        result = self._execute_list()
        assert not_contains_row_with(result.json, {"name": service_name.upper()})

    def describe_should_return_service(self, service_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "object",
                "describe",
                "service",
                f"{self.database}.{self.schema}.{service_name}",
            ],
        )
        assert result.json
        assert result.json[0]["name"] == service_name.upper()  # type: ignore

    def drop_service(self, service_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "object",
                "drop",
                "service",
                f"{self.database}.{self.schema}.{service_name}",
            ],
        )
        assert result.json[0] == {  # type: ignore
            "status": f"{service_name.upper()} successfully dropped."
        }

    def wait_until_service_is_ready(self, service_name: str) -> None:
        self._wait_until_service_reaches_state(service_name, "READY", 900)

    def _wait_until_service_reaches_state(
        self, service_name: str, target_status: Union[str, dict], max_duration: int
    ):
        assert max_duration > 0
        max_counter = math.ceil(max_duration / 10)
        if isinstance(target_status, str):
            target_status = {"status": target_status}
        for i in range(max_counter):
            status = self._execute_status(service_name)
            if contains_row_with(status.json, target_status):
                return
            elif contains_row_with(status.json, {"status": "FAILED"}):
                describe = self._setup.runner.invoke_with_connection_json(
                    ["object", "describe", "service", service_name]
                )
                pytest.fail(
                    dedent(
                        f"""
                    {service_name} service failed before reaching target state:
                    {json.dumps(target_status)}
                    current state:
                    {json.dumps(status)}
                    current describe:
                    {json.dumps(describe)}
                    """
                    )
                )
            time.sleep(10)
        status = self._execute_status(service_name)

        error_message = f"""
{service_name} service didn't reach target state in {max_duration} seconds.
target:
{json.dumps(target_status)}
current:
{json.dumps(status.json)}
"""
        pytest.fail(error_message)

    def _execute_status(self, service_name: str):
        return self._setup.runner.invoke_with_connection_json(
            [
                "spcs",
                "service",
                "status",
                service_name,
                "--database",
                self.database,
                "--schema",
                self.schema,
            ],
        )

    def _execute_list(self):
        return self._setup.runner.invoke_with_connection_json(
            [
                "object",
                "list",
                "service",
            ],
        )

    def _get_spec_path(self):
        return f"{self._setup.test_root_path}/spcs/spec/spec.yml"
