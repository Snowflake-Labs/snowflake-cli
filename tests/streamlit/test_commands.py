import os
import pytest
from tempfile import NamedTemporaryFile
from textwrap import dedent
from unittest import mock
from unittest.mock import call

from click import ClickException

from tests.testing_utils.fixtures import *

STREAMLIT_NAME = "test_streamlit"


@mock.patch("snowflake.connector.connect")
def test_list_streamlit(mock_connector, runner, mock_ctx):
    ctx = mock_ctx()
    mock_connector.return_value = ctx

    result = runner.invoke(["streamlit", "list"])

    assert result.exit_code == 0, result.output
    assert ctx.get_query() == "show streamlits"


@mock.patch("snowflake.connector.connect")
def test_describe_streamlit(mock_connector, runner, mock_ctx):
    ctx = mock_ctx()
    mock_connector.return_value = ctx

    result = runner.invoke(["streamlit", "describe", STREAMLIT_NAME])

    assert result.exit_code == 0, result.output
    assert ctx.get_queries() == [
        f"describe streamlit {STREAMLIT_NAME}",
        f"call SYSTEM$GENERATE_STREAMLIT_URL_FROM_NAME('{STREAMLIT_NAME}')",
    ]


def _put_query(source: str, dest: str):
    return dedent(
        f"put file://{source} {dest} auto_compress=false parallel=4 overwrite=True"
    )


@mock.patch("snowcli.cli.streamlit.commands.typer")
@mock.patch("snowflake.connector.connect")
def test_deploy_streamlit_single_file(
    mock_connector, mock_typer, mock_cursor, runner, mock_ctx
):
    ctx = mock_ctx()
    mock_connector.return_value = ctx

    with NamedTemporaryFile(suffix=".py") as file:
        result = runner.invoke(
            ["streamlit", "deploy", STREAMLIT_NAME, "--file", file.name, "--open"]
        )

    assert result.exit_code == 0, result.output
    assert ctx.get_queries() == [
        "create stage if not exists MOCKDATABASE.MOCKSCHEMA.STREAMLIT",
        _put_query(file.name, "@MOCKDATABASE.MOCKSCHEMA.STREAMLIT/test_streamlit"),
        dedent(
            f"""
    CREATE  STREAMLIT {STREAMLIT_NAME}
    ROOT_LOCATION = '@MOCKDATABASE.MOCKSCHEMA.STREAMLIT/{STREAMLIT_NAME}'
    MAIN_FILE = '{Path(file.name).name}'

    """
        ),
        f"select system$get_snowsight_host()",
    ]

    mock_typer.launch.assert_called_once_with(
        f"https://snowsight.domain/test.region.aws/account/#/streamlit-apps/MOCKDATABASE.MOCKSCHEMA.{STREAMLIT_NAME.upper()}"
    )


@mock.patch("snowflake.connector.connect")
def test_deploy_streamlit_all_files_default_stage(
    mock_connector, mock_cursor, runner, mock_ctx, project_file
):
    ctx = mock_ctx()
    mock_connector.return_value = ctx

    with project_file("example_streamlit") as pdir:
        result = runner.invoke(
            ["streamlit", "deploy", STREAMLIT_NAME, "--file", "main.py"]
        )

    root_path = f"@MOCKDATABASE.MOCKSCHEMA.STREAMLIT/{STREAMLIT_NAME}"
    assert result.exit_code == 0, result.output
    assert ctx.get_queries() == [
        "create stage if not exists MOCKDATABASE.MOCKSCHEMA.STREAMLIT",
        _put_query("main.py", root_path),
        _put_query("environment.yml", root_path),
        _put_query("pages/*", f"{root_path}/pages"),
        dedent(
            f"""
    CREATE  STREAMLIT {STREAMLIT_NAME}
    ROOT_LOCATION = '@MOCKDATABASE.MOCKSCHEMA.STREAMLIT/{STREAMLIT_NAME}'
    MAIN_FILE = 'main.py'

    """
        ),
        f"select system$get_snowsight_host()",
    ]


@mock.patch("snowflake.connector.connect")
def test_deploy_streamlit_all_files_users_stage(
    mock_connector, mock_cursor, runner, mock_ctx, project_file
):
    ctx = mock_ctx()
    mock_connector.return_value = ctx

    with project_file("example_streamlit") as pdir:
        result = runner.invoke(
            [
                "streamlit",
                "deploy",
                STREAMLIT_NAME,
                "--file",
                "main.py",
                "--stage",
                "MY_FANCY_STAGE",
            ]
        )

    root_path = f"@MOCKDATABASE.MOCKSCHEMA.MY_FANCY_STAGE/{STREAMLIT_NAME}"
    assert result.exit_code == 0, result.output
    assert ctx.get_queries() == [
        "create stage if not exists MOCKDATABASE.MOCKSCHEMA.MY_FANCY_STAGE",
        _put_query("main.py", root_path),
        _put_query("environment.yml", root_path),
        _put_query("pages/*", f"{root_path}/pages"),
        dedent(
            f"""
    CREATE  STREAMLIT {STREAMLIT_NAME}
    ROOT_LOCATION = '@MOCKDATABASE.MOCKSCHEMA.MY_FANCY_STAGE/{STREAMLIT_NAME}'
    MAIN_FILE = 'main.py'

    """
        ),
        f"select system$get_snowsight_host()",
    ]


@mock.patch("snowflake.connector.connect")
def test_deploy_streamlit_main_and_environment_files(
    mock_connector, mock_cursor, runner, mock_ctx, project_file
):
    ctx = mock_ctx()
    mock_connector.return_value = ctx

    with project_file("example_streamlit") as pdir:
        (pdir / "pages" / "my_page.py").unlink()
        (pdir / "pages").rmdir()

        result = runner.invoke(
            ["streamlit", "deploy", STREAMLIT_NAME, "--file", "main.py"]
        )

    root_path = f"@MOCKDATABASE.MOCKSCHEMA.STREAMLIT/{STREAMLIT_NAME}"
    assert result.exit_code == 0, result.output
    assert ctx.get_queries() == [
        "create stage if not exists MOCKDATABASE.MOCKSCHEMA.STREAMLIT",
        _put_query("main.py", root_path),
        _put_query("environment.yml", root_path),
        dedent(
            f"""
    CREATE  STREAMLIT {STREAMLIT_NAME}
    ROOT_LOCATION = '@MOCKDATABASE.MOCKSCHEMA.STREAMLIT/{STREAMLIT_NAME}'
    MAIN_FILE = 'main.py'

    """
        ),
        f"select system$get_snowsight_host()",
    ]


@mock.patch("snowflake.connector.connect")
def test_deploy_streamlit_main_and_pages_files(
    mock_connector, mock_cursor, runner, mock_ctx, project_file
):
    ctx = mock_ctx()
    mock_connector.return_value = ctx

    with project_file("example_streamlit") as pdir:
        (pdir / "environment.yml").unlink()
        result = runner.invoke(
            ["streamlit", "deploy", STREAMLIT_NAME, "--file", "main.py"]
        )

    root_path = f"@MOCKDATABASE.MOCKSCHEMA.STREAMLIT/{STREAMLIT_NAME}"
    assert result.exit_code == 0, result.output
    assert ctx.get_queries() == [
        "create stage if not exists MOCKDATABASE.MOCKSCHEMA.STREAMLIT",
        _put_query("main.py", root_path),
        _put_query("pages/*", f"{root_path}/pages"),
        dedent(
            f"""
    CREATE  STREAMLIT {STREAMLIT_NAME}
    ROOT_LOCATION = '@MOCKDATABASE.MOCKSCHEMA.STREAMLIT/{STREAMLIT_NAME}'
    MAIN_FILE = 'main.py'

    """
        ),
        f"call SYSTEM$GENERATE_STREAMLIT_URL_FROM_NAME('{STREAMLIT_NAME}')",
    ]


@pytest.mark.parametrize(
    "opts", [("--pages-dir", "foo/bar"), ("--env-file", "foo.yml")]
)
@mock.patch("snowflake.connector.connect")
def test_deploy_streamlit_nonexisting_file(
    mock_connector, mock_cursor, runner, mock_ctx, project_file, opts
):
    ctx = mock_ctx()
    mock_connector.return_value = ctx

    with project_file("example_streamlit") as pdir:
        result = runner.invoke(
            ["streamlit", "deploy", STREAMLIT_NAME, "--file", "main.py", *opts]
        )

        assert f"Provided file {opts[1]} does not exist" in result.output


@mock.patch("snowflake.connector.connect")
def test_share_streamlit(mock_connector, runner, mock_ctx):
    ctx = mock_ctx()
    mock_connector.return_value = ctx
    role = "other_role"

    result = runner.invoke(["streamlit", "share", STREAMLIT_NAME, role])

    assert result.exit_code == 0, result.output
    assert (
        ctx.get_query() == f"grant usage on streamlit {STREAMLIT_NAME} to role {role}"
    )


@mock.patch("snowflake.connector.connect")
def test_drop_streamlit(mock_connector, runner, mock_ctx):
    ctx = mock_ctx()
    mock_connector.return_value = ctx

    result = runner.invoke(["streamlit", "drop", STREAMLIT_NAME])

    assert result.exit_code == 0, result.output
    assert ctx.get_query() == f"drop streamlit {STREAMLIT_NAME}"
