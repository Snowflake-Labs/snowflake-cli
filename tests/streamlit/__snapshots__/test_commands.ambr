# serializer version: 1
# name: test_artifacts_must_exists
  '''
  +- Error ----------------------------------------------------------------------+
  | During evaluation of DefinitionV20 in project definition following errors    |
  | were encountered:                                                            |
  | For field entities.my_streamlit.streamlit you provided '{'artifacts':        |
  | ['streamlit_app.py', 'foo_bar.py', 'pages/', 'environment.yml'],             |
  | 'main_file': 'streamlit_app.py', 'name': 'test_streamlit_deploy_snowcli',    |
  | 'query_warehouse': 'xsmall', 'stage': 'streamlit', 'title': 'My Fancy        |
  | Streamlit', 'type': 'streamlit'}'. This caused: Value error, Specified       |
  | artifact foo_bar.py does not exist locally.                                  |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_main_file_must_be_in_artifacts
  '''
  +- Error ----------------------------------------------------------------------+
  | During evaluation of DefinitionV20 in project definition following errors    |
  | were encountered:                                                            |
  | For field entities.my_streamlit.streamlit you provided '{'artifacts':        |
  | ['streamlit_app.py', 'utils/utils.py', 'pages/', 'environment.yml'],         |
  | 'main_file': 'foo_bar.py', 'name': 'test_streamlit_deploy_snowcli',          |
  | 'query_warehouse': 'xsmall', 'stage': 'streamlit', 'title': 'My Fancy        |
  | Streamlit', 'type': 'streamlit'}'. This caused: Value error, Specified main  |
  | file foo_bar.py is not included in artifacts.                                |
  +------------------------------------------------------------------------------+
  
  '''
# ---
